### PR TITLE
RTK GPS evaluation 

### PR DESCRIPTION
**Pull request by Annika-wyt**
###   global_ekf.yaml: 
-   Changes:
    -   added **`odom0_pose_rejection_threshold`** and set to 1. If the RTK-GPS measurement is too far away (larger than the threshold) from the current location, the RTK-GPS data will be ignored. If wheel encoder is used, this value can be smaller.
    -   changed **`two_d_mode`** to true to operate the EKF in 2D
    -   changed **`imu0_differential`** to true as the imu always start from 0
    -   changed **`imu0_relative`** to false as the imu always start from 0
-   Performance:
    -   In worse case scenrio, the RTK-GPS drifts up to 40-50cm (even when the covariance for the RTK-GPS signal is small). In best case scenrio, the RTK-GPS can achieve an accuracy of 5-10cm. 
###   navsat.launch:
-   Changes:
    -   added **`broadcast_utm_transform_as_parent_frame`** and 
set to true to add the UTM frame as a parent of map frame.
    -   added **`broadcast_utm_transform`** and set to true to add the UTM frame. This parameter has to be set to true in order to use **`broadcast_utm_transform_as_parent_frame`**. 
-   Notes:
    -   **`yaw_offset`**: This parameters should be set to 0 when the SVEA is facing **EAST**. If not, this parameters should be changed in order to work with RTK-GPS.
### rs_odometry.launch
-   Changes:
    -   remapped the topic **`/set_pose`** to **`/global/set_pose`** to distinguish it from the one for EKF local.

## Outdoor autonomous driving with RTK-GPS
-   An example to demonstrate the performance of the updated outdoor localization stack (with RTK-GPS).
### outdoor_autonomous_driving.launch
-    A launch file that include the localiza.launch, relative_waypoints.py and pure_pursuit_outdoor.py
### relative_waypoints.py
-   A node that take in the GPS location of the points (defined by users) and transform them to the map coordinates based on the initial pose of the robot. These coordinates are then published to **`/outdoor_localization_waypoint`**. 
### pure_pursuit_outdoor.py
-   A node that has a very similar functionalities as the pure_pursuit.py, but it subscribes to **`/outdoor_localization_waypoint`** and use these points to generate the trajectory for pure pursuit. 
### How to run:
```
roslaunch svea_examples outdoor_autonomous_driving.launch is_indoors:=false start_serial:=true device:=/dev/GPS
```
-   Parameters:
    -   **`device`**: the port for the RTK-GPS. If the script is running on SVEA2, the port is fixed to `/dev/GPS`. Otherwise, change it accordingly.
    -   **`use_wheel_encoders`**: true if the script is running on SVEA7.
    -   **`initial_pose_x`**, **`initial_pose_y`**, **`initial_pose_a`**: initial pose of SVEA (x,y,theta)
    -   **`yaw_offset`**: This parameters should be set to 0 when the SVEA is facing **EAST**. If not, this parameters should be changed in order to work with RTK-GPS.
-   Notes:
    -   It might take a few trials before the RTK-GPS connects to the service.
### Performance:
-   The SVEA can trace the path, at least the outline of the path accuractely, given good accuracy of the RTK-GPS. However, since the RTK-GPS drifts, even if the SVEA drives to those points according to rViz, it might not be driving on that exact same place in the physical environment.
-   The first image shows the expected path in orange, and the second image shows the actual path the SVEA took. 
-   The 4 blue points in the images are the inputs for the relative_watpoints.py, and the path is generated using these 4 points. But the entire path drifted to one direction. 

RTK_GPS_Autonomous_driving_expectation
RTK_GPS_Autonomous_driving

### CODE REVIEW

File: get-docker.sh
Comment (kaarmu) [inline get-docker.sh:1]: This might  actually be good to have. Create a new issue named something like "Add get-docker script", write a oneliner about why this is needed and where we can find the script. It should either way not be included in this PR.

File: requirements.txt
Comment (kaarmu) [inline requirements.txt:1]: We should check which are listed as ROS dependencies it might be the case that this is not needed. nats-py should not be here regardless.

File: src/svea_core/src/svea/controllers/pure_pursuit.py
Comment (kaarmu) [inline src/svea_core/src/svea/controllers/pure_pursuit.py:1]: Do not include

File: src/svea_core/src/svea/svea_managers/svea_archetypes.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
 from svea.interfaces import ActuationInterface
 from svea.sensors import Lidar
 from svea.data import BasicDataHandler
+# from svea.park.nonlinear_mpc import NonlinearMPC
```
Comment (kaarmu) [inline src/svea_core/src/svea/svea_managers/svea_archetypes.py:13]: Remove

File: src/svea_examples/launch/outdoor_autonomous_driving.launch
Comment (kaarmu) [inline src/svea_examples/launch/outdoor_autonomous_driving.launch:1]: Name this `outdoor_test.launch`

File: src/svea_examples/scripts/outdoor_examples/relative_waypoints.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+from nav_msgs.msg import Odometry
+from std_msgs.msg import Float64MultiArray
+from visualization_msgs.msg import MarkerArray, Marker
+from tf.transformations import euler_from_quaternion #, quaternion_from_euler
+
+########################################################################################
+###Transform the gps location into map frame (x,y) based on the starting pose of SVEA###
+########################################################################################
+
+class generate():
```
[the remark this answers is not in the document]
Comment (kaarmu) [reply src/svea_examples/scripts/outdoor_examples/relative_waypoints.py:22]: Don't forget in the if statement at the bottom

File: src/svea_sensors/launch/localize.launch [hunk 3]
```diff
@@ -6,74 +6,172 @@
...
     <group if="$(arg use_camera)">
         <param name="use_sim_time" value="false"/>
         <include if="$(eval camera == 'rs')" file="$(find svea_sensors)/launch/rs_odometry.launch">
-            <arg name="xavier" value="$(arg xavier)"/>
+            <arg name="xavier"         value="$(arg xavier)"/>
+            <arg name="is_indoors"     value="$(arg is_indoors)"/>
+            <arg name="initial_pose_x" value="$(arg initial_pose_x)" />
+            <arg name="initial_pose_y" value="$(arg initial_pose_y)" />
+            <arg name="initial_pose_a" value="$(arg initial_pose_a)" />
+            <arg name="pedestrian_demo" value="$(arg pedestrian_demo)"/>
```
Comment (kaarmu) [inline src/svea_sensors/launch/localize.launch:57]: Should not be in here

File: src/svea_sensors/launch/localize.launch [hunk 4]
```diff
@@ -6,74 +6,172 @@
...
+            <param name="baud" value="250000" />
+        </node>
+    </group>
+
     <!-- Start wheel encoders -->
     <group if="$(arg use_wheel_encoders)">
-        <include file="$(find svea_sensors)/launch/wheel_odometry.launch">
-            <arg name="start_serial" value="false"/>
-        </include>
+        <include file="$(find svea_sensors)/launch/wheel_odometry.launch" />
```
Comment (kaarmu) [inline src/svea_sensors/launch/localize.launch:75]: wheel_odometry doesn't start serial by default right?

File: src/svea_sensors/launch/localize.launch [hunk 5]
```diff
@@ -6,74 +6,172 @@
...
-        <param name="publish_pose" value="$(arg publish_pose)"/>
+        <param name="map_frame"             value="$(arg map_frame)"/>
+        <param name="wait_for_transform"    value="$(arg wait_for_transform)"/>
+        <param name="publish_odometry"      value="$(arg publish_odometry)"/>
+        <param name="publish_pose"          value="$(arg publish_pose)"/>
     </node>
 
-</launch>
+    <!-- Use web_camera to detect aruco -->
+    <group if="$(arg use_web_camera)">
```
Comment (kaarmu) [inline src/svea_sensors/launch/localize.launch:164]: Do not include in this PR

File: src/svea_sensors/launch/nats_connector_static.launch
Comment (kaarmu) [inline src/svea_sensors/launch/nats_connector_static.launch:1]: Do not include

File: src/svea_sensors/launch/navsat.launch [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        <param name="yaw_offset" value="$(arg yaw_offset)"/>
+        <param name="zero_altitude" value="true" />
+        <param name="broadcast_utm_transform_as_parent_frame" value="true"/>
+        <param name="broadcast_utm_transform" value="true"/>
+        <remap from="/imu/data" to="/imu/data" />
+        <remap from="/gps/fix" to="/gps/fix" />
+        <remap from="/odometry/filtered" to="/odometry/filtered/global" />
+    </node> 
+
+    <!-- <node pkg="robot_localization" type="navsat_transform_node" name="navsat_transform_node" respawn="true" output="screen">
```
Comment (kaarmu) [inline src/svea_sensors/launch/navsat.launch:20]: remove if not necessary

File: src/svea_sensors/launch/pedestrian_demo.launch
Comment (kaarmu) [inline src/svea_sensors/launch/pedestrian_demo.launch:1]: Do not include

File: src/svea_sensors/launch/transforms.launch [hunk 7]
```diff
@@ -6,16 +6,18 @@
 <launch>
 
     <!-- Launch file arguments -->
-    <arg name="xavier" default="false"/>
-
-
+    <arg name="xavier"          default="false"/>
+    <arg name="use_web_camera"  default="false"/>
+    <arg name="pedestrian_demo" default="false"/>
```
Comment (kaarmu) [inline src/svea_sensors/launch/transforms.launch:11]: Do not include

File: src/svea_sensors/launch/transforms.launch [hunk 8]
```diff
@@ -24,6 +26,10 @@
         <node pkg="tf2_ros" type="static_transform_publisher" name="lidarlink_broadcaster"  args="0.284 0 0.2 0 0 0 1 base_link laser" />
         <node pkg="tf2_ros" type="static_transform_publisher" name="t265_link_broadcaster"  args="0.09 0 0.215 0 -1.57 3.14 base_link t265_camera_pose_frame" />
         <node pkg="tf2_ros" type="static_transform_publisher" name="t265_odom_broadcaster"  args="0.09 0 0.215 0 -1.57 3.14 odom t265_camera_odom_frame" />
+        <node pkg="tf2_ros" type="static_transform_publisher" name="gps_broadcaster"        args="0.215 0 0.215 0 0 0 base_link gps" />
+        <node pkg="tf2_ros" type="static_transform_publisher" name="webcamlink_broadcaster" args="0.345 0 0.250 -1.57 0 -1.57 base_link camera" />
```
Comment (kaarmu) [inline src/svea_sensors/launch/transforms.launch:30]: Do not include the last webcam/aruco stuff

File: src/svea_sensors/launch/wheel_odometry.launch
Comment (kaarmu) [inline src/svea_sensors/launch/wheel_odometry.launch:1]: Okay this looks good

File: src/svea_sensors/package.xml [hunk 9]
```diff
@@ -31,8 +31,12 @@
   <depend>rosserial_msgs</depend>
   <depend>rosserial_python</depend>
   <depend>svea_msgs</depend>
+  <depend>aruco_msgs</depend>
```
Comment (kaarmu) [inline src/svea_sensors/package.xml:34]: No webcam/aruco stuff. Only keep deps for rtk

File: src/svea_sensors/params/robot_localization/global_ekf_static.yaml
[the remark this answers is not in the document]
Comment (kaarmu) [reply src/svea_sensors/params/robot_localization/global_ekf_static.yaml:1]: if you haven't done so already

File: src/svea_sensors/params/robot_localization/rs_ekf_static.yaml
Comment (kaarmu) [inline src/svea_sensors/params/robot_localization/rs_ekf_static.yaml:1]: Comment on difference between rs_ekf and rs_ekf_static

File: src/svea_sensors/params/robot_localization/usbc_web_camera.yaml
Comment (kaarmu) [inline src/svea_sensors/params/robot_localization/usbc_web_camera.yaml:1]: not in this PR

File: src/svea_sensors/params/robot_localization/web_camera.yaml
Comment (kaarmu) [inline src/svea_sensors/params/robot_localization/web_camera.yaml:1]: not in this PR

File: src/svea_sensors/scripts/aruco_detect.py
Comment (kaarmu) [inline src/svea_sensors/scripts/aruco_detect.py:1]: not included in this PR

File: test.py
Comment (kaarmu) [inline test.py:1]: do not include

**Review by kaarmu — CHANGES_REQUESTED** [review]
Nice work! I've left some comments throughout what to save for later. At places I also asked for an explanation of why it should be included. You can just answer with a small comment as a reply to the comment.

:warning: Also, before you start resolving any of the listed notes in this review you should create a new branch to save your setup. This is because of the second Aruco PR later on.

Beside the comments listed throughout, here are some other overarching changes you should do.
- `rsu_msgs` should not be included
-  What is `svea_core/src/svea/park` for?

**Comment by kaarmu** [comment]
Is there something that can be done to `yaw_offset`? Feels like a hassle to always think about cardinal direction

File: src/svea_sensors/launch/localize.launch [hunk 10]
```diff
@@ -6,74 +6,172 @@
...
+    <arg name="initial_pose_a"          default="0.0" />
+
+    <arg name="use_wheel_encoders"      default="false" />
+    <arg name="is_indoors"              default="true" />
+    <!-- Should be set to true, if only the localization is running (i.e. this is needed in order to have wheel encoders
+    and actuation_to_twist working) -->
+    <arg name="start_serial"            default="false" />
+    <arg name="device"                  default="/dev/GPS" />
+    <arg name="delay"                   default="2.0"/>
+    <!-- yaw_offset is set to 0 if the robot is facing East during initialization -->
```
[the remark this answers is not in the document]
Comment (Annika-wyt) [reply src/svea_sensors/launch/localize.launch:29]: The documentation for robot localization mentioned that the navsat node assumes that all heading data starts with its 0 points facing east. It is for aligning the GPS signal with the imu.

File: src/svea_sensors/launch/localize.launch (hunk 4 above)
Comment (Annika-wyt) [reply src/svea_sensors/launch/localize.launch:75]: It does not.

File: src/svea_sensors/params/robot_localization/global_ekf_static.yaml
[the remark this answers is not in the document]
Comment (Annika-wyt) [reply src/svea_sensors/params/robot_localization/global_ekf_static.yaml:1]: I think I should not include this file for this pr. It is for the ArUco stuff.

File: src/svea_sensors/params/robot_localization/rs_ekf_static.yaml
Comment (Annika-wyt) [reply src/svea_sensors/params/robot_localization/rs_ekf_static.yaml:1]: Same as global_ekf_static, it is for the ArUco stuff

File: src/svea_sensors/scripts/actuation_to_twist.py [hunk 11]
```diff
@@ -149,7 +149,8 @@ def calc_current_velocity(self):
 
     def _sim_esc(self, velocity, target_velocity):
         # simulates esc dynamics
-        tau = self.tau1 if self._actuation_values.velocity else self.tau0
+        #!!!!!!tau = self.tau1 if self._actuation_values.velocity else self.tau0
```
[the remark this answers is not in the document]
Comment (Annika-wyt) [reply src/svea_sensors/scripts/actuation_to_twist.py:152]: The tau should be dependent on the gear instead of the velocity? At least this is how i understand it

File: src/svea_sensors/scripts/actuation_to_twist.py [hunk 12]
```diff
@@ -137,7 +137,7 @@ def calc_current_velocity(self):
         if self._last_calc_time is not None:
             dt = (time_now - self._last_calc_time).to_sec()
         else:
-            dt = 0.1
+            dt = 0.01 #0.1
```
[the remark this answers is not in the document]
Comment (Annika-wyt) [reply src/svea_sensors/scripts/actuation_to_twist.py:140]: I think when I changed it, I compared it with the other dt, and seems to be closer to 0.01. But the else statement is for the first step only, so I don't think it matters much in that sense.

File: src/svea_sensors/scripts/odom_pub.py
[the remark this answers is not in the document]
Comment (Annika-wyt) [reply src/svea_sensors/scripts/odom_pub.py:1]: I think it is one of the files written by the Italians

File: src/svea_sensors/scripts/odometry_corrected.py
[the remark this answers is not in the document]
Comment (Annika-wyt) [reply src/svea_sensors/scripts/odometry_corrected.py:1]: Same for this file, is also from the Italians. I don't think these files are used, but I have to double-check tomorrow.

File: src/svea_sensors/scripts/static_svea_pose.py
[the remark this answers is not in the document]
Comment (Annika-wyt) [reply src/svea_sensors/scripts/static_svea_pose.py:1]: Should be one of the old scripts I wrote for the ArUco stuff. I will delete it.

File: src/svea_sensors/scripts/tf_listener.py
[the remark this answers is not in the document]
Comment (Annika-wyt) [reply src/svea_sensors/scripts/tf_listener.py:1]: It is from the Italians. I think they used it to test the GPS stuff, will check again tomorrow to see if it is in use.

**Comment by Annika-wyt** [comment]
> Is there something that can be done to `yaw_offset`? Feels like a hassle to always think about cardinal direction

I think I can check if we can change the way the IMU publishes the data. Based on the description from the Robot localization package, there should be some kind of standard for the IMU to either report 0 when facing north or east.

File: requirements.txt
Comment (kaarmu) [inline requirements.txt:1]: All of these are needed?

**Comment by Annika-wyt** [comment]
> > Is there something that can be done to `yaw_offset`? Feels like a hassle to always think about cardinal direction
> 
> I think I can check if we can change the way the IMU publishes the data. Based on the description from the Robot localization package, there should be some kind of standard for the IMU to either report 0 when facing north or east.

Conclusion: 
After searching on the internet, there is default mapping for each IMU on cardinal direction. For ours, 0 -> North (https://www.bosch-sensortec.com/media/boschsensortec/downloads/application_notes_1/bst-bno055-an007.pdf Page4). So, for the yaw_offset, it should be set to around 1.57. However, **this does not help** with the starting orientation of the robot. Which means, whenever we start the robot, we will have to measure its heading (yaw) and feed into the launch file as the initial_pose_a argument.

File: src/svea_examples/scripts/pure_pursuit.py
Comment (kaarmu) [inline src/svea_examples/scripts/pure_pursuit.py:1]: This one should not have been deleted?

File: src/svea_sensors/launch/navsat.launch [hunk 13]
```diff
@@ -1,14 +1,13 @@
...
     <arg name="delay"       default="2.0" />
-    <arg name="yaw_offset"  default="0"/>
 
     <node pkg="robot_localization" type="navsat_transform_node" name="navsat_transform_node" respawn="true" output="screen">
         <param name="publish_filtered_gps" value="true"/>
         <param name="wait_for_datum" value="false"/>
         <param name="delay" value="$(arg delay)" />
         <param name="magnetic_declination_radians" value="0.1460841"/>
-        <param name="yaw_offset" value="$(arg yaw_offset)"/>
+        <param name="yaw_offset" value="1.5707963"/>
```
Comment (kaarmu) [inline src/svea_sensors/launch/navsat.launch:10]: This shouldn't be hard coded or...?

File: src/svea_sensors/launch/transforms.launch (hunk 8 above)
Comment (kaarmu) [reply src/svea_sensors/launch/transforms.launch:28]: There is still a webcamlink_broadcaster here

**Review by kaarmu — CHANGES_REQUESTED** [review]
Nice work. A few more changes to do but it's starting to take form

File: src/svea_sensors/launch/navsat.launch (hunk 13 above)
Comment (Annika-wyt) [reply src/svea_sensors/launch/navsat.launch:10]: The yaw_offset is based on the documentation of the IMU, saying the mapping for the cardinal directions. So, it should be hard coded. The parameter we should be changing during initialization is initial_pose_a, which define the starting yaw for the SVEA.

File: requirements.txt
Comment (Annika-wyt) [reply requirements.txt:1]: pyserial - needed in rtk.manager
pyrtcm - is imported to rtk.manger, but it seems like it is not used. I think this can be removed.
geopy - needed in relative_waypoints.py

**Review by kaarmu — COMMENTED** [review]
I don't have any more specific comments codewise. I saw you removed account credentials. That's good, make sure there aren't any else.

I will approve it tomorrow when we're in the lab

File: src/svea_sensors/package.xml
Comment (kaarmu) [inline src/svea_sensors/package.xml:1]: Do not remove

File: src/svea_examples/scripts/pure_pursuit.py
Comment (kaarmu) [inline src/svea_examples/scripts/pure_pursuit.py:1]: Make executable

File: src/svea_sensors/params/rviz/navsat.rviz
Comment (kaarmu) [inline src/svea_sensors/params/rviz/navsat.rviz:1]: Either
- move to svea_core/rviz
- move to svea_sensors/rviz
- remove

Keep only one navsat.rviz (currently there is two, one in svea_sensors and one in svea_core)

File: src/svea_core/src/svea/controllers/pure_pursuit.py
Comment (kaarmu) [inline src/svea_core/src/svea/controllers/pure_pursuit.py:1]: Revert to original version

File: src/svea_core/src/svea/park/drive_state.py
Comment (kaarmu) [inline src/svea_core/src/svea/park/drive_state.py:1]: Do not include in this PR. We can make another PR for MPC support

Comment (kaarmu) [reply src/svea_core/src/svea/park/drive_state.py:1]: The entire park directory, that is

File: src/svea_sensors/CMakeLists.txt
Comment (kaarmu) [inline src/svea_sensors/CMakeLists.txt:1]: Remove aruco stuff (comments also)

File: src/svea_sensors/scripts/odom_pub.py
Comment (kaarmu) [inline src/svea_sensors/scripts/odom_pub.py:1]: Check if used

Comment (kaarmu) [reply src/svea_sensors/scripts/odom_pub.py:1]: remove otherwise

File: src/svea_sensors/scripts/odometry_corrected.py
Comment (kaarmu) [inline src/svea_sensors/scripts/odometry_corrected.py:1]: Check if used, remove otherwise

File: src/svea_sensors/scripts/tf_listener.py
Comment (kaarmu) [inline src/svea_sensors/scripts/tf_listener.py:1]: Check if used, remove otherwise

Comment (kaarmu) [reply src/svea_sensors/scripts/tf_listener.py:1]: If it is described in README, keep it. Same with other ones

**Review by kaarmu — CHANGES_REQUESTED** [review]
Fix this, then approve as we discussed

File: .gitignore
Comment (kaarmu) [inline .gitignore:1]: .gitignore should been deleted

File: src/svea_sensors/params/rviz/navsat.rviz
Comment (kaarmu) [inline src/svea_sensors/params/rviz/navsat.rviz:1]: Move to svea_sensors/rviz/navsat.rviz

**Review by kaarmu — COMMENTED** [review]
2 more comments... haha

**Review by kaarmu — APPROVED** [review]
Done! :1st_place_medal:

**Comment by kaarmu** [comment]
You can see in the box that there are a lot of conflicting files. You need to resolve them...
